### PR TITLE
Add ordered_prefetch

### PR DESCRIPTION
**Pull request by Algue-Rythme**
Here is a possible implementation of the `Buffer.ordered_prefetch` method.

The `OrderedPrefetch` is a `Stream` produced by prefetching in parallel (using a `ThreadPool`) contiguous elements from a `Buffer`. The implementation is an hybrid between `Buffer.to_stream` and `Stream.prefetch`.

`Prefetch` fetches the elements of the queue `prefetchCache_` in order, but the order in which these threads call `Stream.next` is not guaranteed. `OrderedPrefetch` also fetches the elements of a vector `prefetchCache_` in order, but each package task enqueued in the `pool_` is assigned a single `idx` captured by the lambda, that points to an index in the underlying `buffer_`. A `std::lock_guard` is used to ensure atomicity of calls to `OrderedPrefetch .next`. 

Interestingly, the main limitation of this approach is its lack of expressiveness: it must replaces a `Buffer.to_stream` call, but it is not necessarily slower than the original `Prefetch`. However I documented it as a slower counterpart.   

**TODO**:

- [x] Perform a benchmark

Feedbacks are welcome.

### CODE REVIEW

File: python/src/wrap_stream.cpp [hunk 1]
```diff
@@ -341,6 +341,10 @@ void init_mlx_data_stream(py::module& m) {
                 Prefetch can be used both to parallelize operations but also to
                 overlap computation with data loading in a background thread.
 
+                This prefetching order is not deterministic and samples ordering depends
```
Comment (angeloskath) [inline python/src/wrap_stream.cpp:344]:
```suggestion
                This prefetching order is not deterministic and samples' ordering depends
```

File: docs/src/_static/mlx_logo.png
Comment (angeloskath) [inline docs/src/_static/mlx_logo.png:1]: Not sure what is going on with the two images. I would probably check them out from main.

File: mlx/data/stream/OrderedPrefetch.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  if (idx < 0) {
+    return Sample();
+  } else {
+    int f_idx = idx % prefetchSize_;
+    std::future<Sample> fsample(std::move(prefetchCache_[f_idx]));
+    prefetchCache_[f_idx] =
+        pool_->enqueue([b = buffer_, idx] { return b->get(idx); });
+    return fsample.get();
+  }
```
Comment (angeloskath) [inline mlx/data/stream/OrderedPrefetch.cpp:53]: All the above is really great. I would probably unlock before calling `fsample.get()`. It probably won't make a difference in performance but there is no real reason to keep everything locked while waiting for `fsample.get()`.

**Review by angeloskath — CHANGES_REQUESTED** [review]
This looks awesome thanks for adding it!

I left a couple of comments, let me know what you think and then we can merge.

**Comment by Algue-Rythme** [comment]
Thanks for the feedback. 

I call `unlock()` early per your advice.  

I added two unit test, including a corner case.

### Benchmark

I added a line to the only benchmark that starts from a Buffer, i.e `caltech101`.

On my M3 with `prefetch`:

```
Benchmark MLX Caltech 101

iterate_no_workers       4.351 s       62.521 samples/s      3 runs     
iterate_workers_8        0.683 s      401.271 samples/s      3 runs     
iterate_workers_16       0.445 s      620.840 samples/s      3 runs
```

My `ordered_prefetch` on my M3:

```
Benchmark MLX Caltech 101

iterate_no_workers       4.422 s       61.515 samples/s      3 runs     
iterate_workers_8        0.684 s      397.822 samples/s      3 runs     
iterate_workers_16       0.443 s      613.973 samples/s      3 runs
```

It is about the same speed, which is consistent with what we would expect. Note: the whole pipeline has been replaced from `to_stream -> processing -> prefetch` to `processing -> ordered_prefetch` which means that the `Buffer` version of each processing has been called in place of the `Stream` version.

**Review by angeloskath — APPROVED** [review]
Thanks for the awesome addition!
